### PR TITLE
XCOM-281: Configuration now includes JWT issuer and expiration date

### PR DESCRIPTION
**Pull request by Nickersoft**
@clintonb @jimabramson @feanil

### CODE REVIEW

File: playbooks/roles/ecommerce/defaults/main.yml [hunk 1]
```diff
@@ -42,9 +42,10 @@ ECOMMERCE_EDX_API_KEY: 'put-your-edx-api-auth-token-here'
 ECOMMERCE_ECOMMERCE_URL_ROOT: 'https://www.example.com'
 ECOMMERCE_LMS_URL_ROOT: 'https://www.example.com'
 ECOMMERCE_JWT_SECRET_KEY: 'generated-key-that-matches-ECOMMERCE_API_SIGNING_KEY-in-edxapp'
+ECOMMERCE_JWT_ISSUER: 'https://{{ ECOMMERCE_LMS_URL_ROOT }}/oauth2'
```
Comment (clintonb) [inline playbooks/roles/ecommerce/defaults/main.yml]: This should be `'{{ ECOMMERCE_JWT_ISSUER }}'`. The real value will come from the secure repo, configured by @feanil.

**Comment by Nickersoft** [comment]
@clintonb Updated

File: playbooks/roles/edxapp/defaults/main.yml [hunk 2]
```diff
@@ -976,3 +976,9 @@ edxapp_cms_variant: cms
 
 # Worker Settings
 worker_django_settings_module: 'aws'
+
+# JWT Settings
+JWT_AUTH:
```
Comment (clintonb) [inline playbooks/roles/edxapp/defaults/main.yml]: This dict shouldn't exist. You only created two settings, neither of which is contained in this dict.

File: playbooks/roles/edxapp/defaults/main.yml [hunk 3]
```diff
@@ -976,3 +976,9 @@ edxapp_cms_variant: cms
...
+    JWT_SECRET_KEY: '{{ EDXAPP_ECOMMERCE_API_SIGNING_KEY }}'
+    JWT_ISSUER: 'https://{{ ECOMMERCE_JWT_ISSUER }}/oauth2'
```
Comment (clintonb) [inline playbooks/roles/edxapp/defaults/main.yml]: As above, this should just be JWT_ISSUER.

File: playbooks/roles/ecommerce/defaults/main.yml [hunk 4]
```diff
@@ -42,9 +42,10 @@ ECOMMERCE_EDX_API_KEY: 'put-your-edx-api-auth-token-here'
 ECOMMERCE_ECOMMERCE_URL_ROOT: 'https://www.example.com'
 ECOMMERCE_LMS_URL_ROOT: 'https://www.example.com'
 ECOMMERCE_JWT_SECRET_KEY: 'generated-key-that-matches-ECOMMERCE_API_SIGNING_KEY-in-edxapp'
+ECOMMERCE_JWT_ISSUER: 'https://{{ ECOMMERCE_JWT_ISSUER }}/oauth2'
```
Comment (jimabramson) [inline playbooks/roles/ecommerce/defaults/main.yml]: if JWT_ISSUER is really the hostname of the issuer I think the name should change to JWT_ISSUER_HOST

File: playbooks/roles/ecommerce/defaults/main.yml (hunk 1 above)
Comment (Nickersoft) [inline playbooks/roles/ecommerce/defaults/main.yml]: Fixed

File: playbooks/roles/edxapp/defaults/main.yml (hunk 2 above)
Comment (Nickersoft) [inline playbooks/roles/edxapp/defaults/main.yml]: Fixed

File: playbooks/roles/edxapp/defaults/main.yml (hunk 3 above)
Comment (Nickersoft) [inline playbooks/roles/edxapp/defaults/main.yml]: Fixed

File: playbooks/roles/ecommerce/defaults/main.yml (hunk 4 above)
Comment (Nickersoft) [inline playbooks/roles/ecommerce/defaults/main.yml]: Fixed

**Comment by Nickersoft** [comment]
@jimabramson @clintonb @feanil Updated

File: playbooks/roles/edxapp/defaults/main.yml [hunk 5]
```diff
@@ -434,6 +437,10 @@ EDXAPP_CELERY_WORKERS:
 EDXAPP_DEFAULT_CACHE_VERSION: "1"
 EDXAPP_OAUTH_ENFORCE_SECURE: True
 
+# JWT Settings
+JWT_ISSUER: {{ EDXAPP_JWT_ISSUER }}
```
Comment (clintonb) [inline playbooks/roles/edxapp/defaults/main.yml]: Move both of these declarations to be nested within generic_env_config below.

Comment (clintonb) [inline playbooks/roles/edxapp/defaults/main.yml]: Also, correction from my original comment. You can compose the URL as we do for OAUTH_OIDC_ISSUER: https://github.com/edx/configuration/blob/b8ee65fa98638608a9b7398c448c5abee07c23ec/playbooks/roles/edxapp/defaults/main.yml#L682.

Comment (Nickersoft) [inline playbooks/roles/edxapp/defaults/main.yml]: @clintonb Fixed in ecommerce settings. Should the same be done for edxapp as well?

**Comment by Nickersoft** [comment]
@clintonb Updated

File: playbooks/roles/ecommerce/defaults/main.yml [hunk 6]
```diff
@@ -103,7 +105,9 @@ ECOMMERCE_SERVICE_CONFIG:
   COMMERCE_API_URL: '{{ ECOMMERCE_LMS_URL_ROOT }}/api/commerce/v1/'
   LMS_DASHBOARD_URL: '{{ ECOMMERCE_LMS_URL_ROOT }}/dashboard'
   JWT_AUTH:
-    JWT_SECRET_KEY: '{{ ECOMMERCE_JWT_SECRET_KEY }}'
+    JWT_SECRET_KEY: {{ ECOMMERCE_JWT_SECRET_KEY }}
+    JWT_ISSUER: 'https://{{ ECOMMERCE_JWT_ISSUER }}/oauth2'
```
Comment (clintonb) [inline playbooks/roles/ecommerce/defaults/main.yml]: This should only be ECOMMERCE_JWT_ISSUER. Composition happens above when declaring a default.

File: playbooks/roles/ecommerce/defaults/main.yml [hunk 7]
```diff
@@ -42,9 +42,11 @@ ECOMMERCE_EDX_API_KEY: 'put-your-edx-api-auth-token-here'
 ECOMMERCE_ECOMMERCE_URL_ROOT: 'https://www.example.com'
 ECOMMERCE_LMS_URL_ROOT: 'https://www.example.com'
 ECOMMERCE_JWT_SECRET_KEY: 'generated-key-that-matches-ECOMMERCE_API_SIGNING_KEY-in-edxapp'
+ECOMMERCE_JWT_ISSUER: 'https://www.example.com'
```
Comment (clintonb) [inline playbooks/roles/ecommerce/defaults/main.yml]: '{{ ECOMMERCE_LMS_URL_ROOT }}/oauth2'

File: playbooks/roles/ecommerce/defaults/main.yml (hunk 6 above)
Comment (clintonb) [inline playbooks/roles/ecommerce/defaults/main.yml]: Correction: I see what we are doing for other URLs. This should be '{{ ECOMMERCE_LMS_URL_ROOT }}/oauth2'

File: playbooks/roles/ecommerce/defaults/main.yml (hunk 7 above)
Comment (clintonb) [inline playbooks/roles/ecommerce/defaults/main.yml]: This can be removed.

File: playbooks/roles/edxapp/defaults/main.yml [hunk 8]
```diff
@@ -717,6 +720,8 @@ generic_env_config:  &edxapp_generic_env
   WIKI_ENABLED: true
   SYSLOG_SERVER:  "{{ EDXAPP_SYSLOG_SERVER }}"
   LOG_DIR:  "{{ COMMON_DATA_DIR }}/logs/edx"
+  JWT_ISSUER: {{ EDXAPP_JWT_ISSUER }}
```
Comment (clintonb) [inline playbooks/roles/edxapp/defaults/main.yml]: Further proof you shouldn't listen to me. Copy the behavior of OAUTH_OIDC_ISSUER: "https://{{ EDXAPP_LMS_BASE }}/oauth2"

File: playbooks/roles/ecommerce/defaults/main.yml [hunk 9]
```diff
@@ -103,7 +105,9 @@ ECOMMERCE_SERVICE_CONFIG:
   COMMERCE_API_URL: '{{ ECOMMERCE_LMS_URL_ROOT }}/api/commerce/v1/'
   LMS_DASHBOARD_URL: '{{ ECOMMERCE_LMS_URL_ROOT }}/dashboard'
   JWT_AUTH:
-    JWT_SECRET_KEY: '{{ ECOMMERCE_JWT_SECRET_KEY }}'
+    JWT_SECRET_KEY: {{ ECOMMERCE_JWT_SECRET_KEY }}
```
Comment (feanil) [inline playbooks/roles/ecommerce/defaults/main.yml]: This should be quoted for Ansible to render this correctly.

File: playbooks/roles/ecommerce/defaults/main.yml (hunk 6 above)
Comment (feanil) [inline playbooks/roles/ecommerce/defaults/main.yml]: You have the protocol in front of `ECOMMERCE_JWT_ISSUER` but it is also part of the base value of `ECOMMERCE_JWT_ISSUER`

Comment (Nickersoft) [inline playbooks/roles/ecommerce/defaults/main.yml]: Fixed

File: playbooks/roles/ecommerce/defaults/main.yml (hunk 7 above)
Comment (Nickersoft) [inline playbooks/roles/ecommerce/defaults/main.yml]: Fixed

File: playbooks/roles/edxapp/defaults/main.yml (hunk 8 above)
Comment (Nickersoft) [inline playbooks/roles/edxapp/defaults/main.yml]: Fixed

File: playbooks/roles/ecommerce/defaults/main.yml (hunk 9 above)
Comment (Nickersoft) [inline playbooks/roles/ecommerce/defaults/main.yml]: Fixed

File: playbooks/roles/edxapp/defaults/main.yml [hunk 10]
```diff
@@ -230,6 +230,9 @@ EDXAPP_BULK_EMAIL_LOG_SENT_EMAILS: false
 EDXAPP_UNIVERSITY_EMAIL: 'university@example.com'
 EDXAPP_PRESS_EMAIL: 'press@example.com'
 
+EDXAPP_JWT_ISSUER: 'https://www.example.com'
```
Comment (clintonb) [inline playbooks/roles/edxapp/defaults/main.yml]: This should be removed.

**Comment by Nickersoft** [comment]
@clintonb Updated

**Comment by clintonb** [comment]
Looks good to me. @feanil any feedback?

File: playbooks/roles/edxapp/defaults/main.yml [hunk 11]
```diff
@@ -229,6 +229,7 @@ EDXAPP_BULK_EMAIL_DEFAULT_FROM_EMAIL: 'no-reply@example.com'
 EDXAPP_BULK_EMAIL_LOG_SENT_EMAILS: false
 EDXAPP_UNIVERSITY_EMAIL: 'university@example.com'
 EDXAPP_PRESS_EMAIL: 'press@example.com'
+EDXAPP_JWT_EXPIRATION: 30
```
Comment (feanil) [inline playbooks/roles/edxapp/defaults/main.yml]: Can we add a comment to indicate what the unit of this value is?

**Comment by feanil** [comment]
After adding a comment about units :+1:

**Comment by Nickersoft** [comment]
@clintonb, any further comments or is it good to merge?

**Comment by clintonb** [comment]
:+1:

**Pull request merged**